### PR TITLE
Shared tasks usage at the pool level for worker-threads

### PR DESCRIPTION
**Pull request by jerome-benoit**
## Summary 

Shared array buffer usage for a shared memory like between workers in the pool is only relevant for worker-threads. 
So the implementation has to: 
+ use the tasks usage map for cluster pool
+ use shared array buffer for worker-threads pool 

An abstraction for tasks usage statistics should be considered, but since their internal representation between pool type is so different, it might not be a clean solution for the implementation. 

## PR Checklist

- [x] Please add a description of your changes.
- [ ] We need your changes to come with unit tests in order to keep this project in quality and easy to maintain.
- [ ] If your changes contain any new feature please be sure to document it.
- [ ] Please add a link to the open issue or task that this pull request will resolve.
- [ ] Add a note to the changelog to indicate the change.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

If yes please indicate it and migration info into the CHANGELOG.md file.

---

Signed-off-by: Jérôme Benoit <jerome.benoit@sap.com>

### CODE REVIEW

**Comment by Shinigami92** [comment]
Will look into this in the next days

**Review by pioardi — APPROVED** [review]
Looks good for me, thanks @jerome-benoit

**Comment by Shinigami92** [comment]
@jerome-benoit If you feel comfortable with this, please feel free to merge and bring the project forward from your side

As you have already potential recoginzed, I don't have so much time for poolifier at all.
To much in work, vite and my own plugins and then my own rest of free time

**Comment by jerome-benoit** [comment]
> @jerome-benoit If you feel comfortable with this, please feel free to merge and bring the project forward from your side

That patches stack is not stable enough on my side for a merge. I have issue with the accurate statistics introduction and its impact on worker selection strategies. I'm still working on it but that's not trivial to make it work. I'm also investigating for an intermediate solution with less accurate stats and some disabled new strategies to permit to merge that work without breaking everything. 

> As you have already potential recognized, I don't have so much time for poolifier at all.
> To much in work, vite and my own plugins and then my own rest of free time

No pb.

**Comment by jerome-benoit** [comment]
> @jerome-benoit If you feel comfortable with this, please feel free to merge and bring the project forward from your side

I've been a bit busy lately in my private life and work, but I've still been able to make slow progress on that patches set. I think I'm very close to have something working, hopefully for next year.

**Comment by pioardi** [comment]
@jerome-benoit  hei there, do you think I can review this PR ?

**Comment by jerome-benoit** [comment]
> @jerome-benoit hei there, do you think I can review this PR ?

The PRs stack for that feature is not in a mergeable state. And I still do not have time to work on them to fix the remaining far from trivial issues.

File: src/pools/abstract-pool.ts [hunk 1]
```diff
@@ -377,4 +406,78 @@ export abstract class AbstractPool<
       this.emitter?.emit('busy')
     }
   }
+
+  /**
+   * Increase the number of tasks that the given worker has applied.
+   *
+   * @param worker Worker which running tasks is increased.
+   */
+  protected increaseWorkerRunningTasks (worker: Worker): void {
```
Comment (jerome-benoit) [inline src/pools/abstract-pool.ts:415]:
```suggestion
  private increaseWorkerRunningTasks (worker: Worker): void {
```

File: src/pools/abstract-pool.ts [hunk 2]
```diff
@@ -377,4 +406,78 @@ export abstract class AbstractPool<
...
+    this.stepWorkerRunningTasks(worker, 1)
+  }
+
+  /**
+   * Decrease the number of tasks that the given worker has applied.
+   *
+   * @param worker Worker which running tasks is decreased.
+   */
+  protected decreaseWorkerRunningTasks (worker: Worker): void {
+    this.stepWorkerRunningTasks(worker, 1)
```
Comment (jerome-benoit) [inline src/pools/abstract-pool.ts:425]:
```suggestion
    this.stepWorkerRunningTasks(worker, -1)
```

File: src/pools/abstract-pool.ts [hunk 3]
```diff
@@ -377,4 +406,78 @@ export abstract class AbstractPool<
...
+  private increaseWorkerRunningTasks (worker: Worker): void {
+    this.stepWorkerRunningTasks(worker, 1)
+  }
+
+  /**
+   * Decrease the number of tasks that the given worker has applied.
+   *
+   * @param worker Worker which running tasks is decreased.
+   */
+  protected decreaseWorkerRunningTasks (worker: Worker): void {
```
Comment (jerome-benoit) [inline src/pools/abstract-pool.ts:424]:
```suggestion
  private decreaseWorkerRunningTasks (worker: Worker): void {
```

File: tests/worker/abstract-worker.test.js [hunk 4]
```diff
@@ -20,17 +20,20 @@ describe('Abstract worker test suite', () => {
...
+    expect(worker.opts.usage).toBe(false)
   })
 
   it('Verify that worker options are set at worker creation', () => {
     const worker = new ClusterWorker(() => {}, {
       maxInactiveTime: 6000,
       async: true,
-      killBehavior: KillBehaviors.HARD
+      killBehavior: KillBehaviors.HARD,
+      usage: true
```
Comment (jerome-benoit) [inline tests/worker/abstract-worker.test.js:31]:
```suggestion
```

File: tests/worker/abstract-worker.test.js [hunk 5]
```diff
@@ -20,17 +20,19 @@ describe('Abstract worker test suite', () => {
...
     expect(worker.opts.async).toBe(false)
+    expect(worker.opts.usage).toBe(false)
   })
 
   it('Verify that worker options are set at worker creation', () => {
     const worker = new ClusterWorker(() => {}, {
       maxInactiveTime: 6000,
       async: true,
-      killBehavior: KillBehaviors.HARD
+      killBehavior: KillBehaviors.HARD,
```
Comment (jerome-benoit) [inline tests/worker/abstract-worker.test.js:30]:
```suggestion
      killBehavior: KillBehaviors.HARD
```

File: tests/worker/abstract-worker.test.js [hunk 6]
```diff
@@ -31,6 +32,7 @@ describe('Abstract worker test suite', () => {
     expect(worker.opts.maxInactiveTime).toBe(6000)
     expect(worker.opts.killBehavior).toBe(KillBehaviors.HARD)
     expect(worker.opts.async).toBe(true)
+    expect(worker.opts.usage).toBe(true)
```
Comment (jerome-benoit) [inline tests/worker/abstract-worker.test.js:35]:
```suggestion
```

File: tests/worker/abstract-worker.test.js [hunk 7]
```diff
@@ -20,6 +20,7 @@ describe('Abstract worker test suite', () => {
     expect(worker.opts.maxInactiveTime).toBe(1000 * 60)
     expect(worker.opts.killBehavior).toBe(KillBehaviors.SOFT)
     expect(worker.opts.async).toBe(false)
+    expect(worker.opts.usage).toBe(false)
```
Comment (jerome-benoit) [inline tests/worker/abstract-worker.test.js:23]:
```suggestion
```

File: tests/pools/abstract/abstract-pool.test.js [hunk 8]
```diff
@@ -13,7 +13,8 @@ describe('Abstract pool test suite', () => {
   )
   class StubPoolWithWorkerTasksUsageMapClear extends FixedThreadPool {
     removeAllWorker () {
-      this.workersTasksUsage.clear()
+      this.workers.length = 0
```
Comment (jerome-benoit) [inline tests/pools/abstract/abstract-pool.test.js:16]:
```suggestion
      this.workers = []
```

**Comment by jerome-benoit** [comment]
The added complexity with very limited gain since 2.4.x does not worth the effort

**Pull request closed without merging**